### ISSUE TITLE
Support for Storage-attached indexing (SAI)

### ISSUE THREAD

**Issue by Saghetti0** [scylladb/scylladb#19999]
[Cassandra 5.0](https://cassandra.apache.org/_/Apache-Cassandra-5.0-Moving-Toward-an-AI-Driven-Future.html) introduces [storage-attached indexing](https://cassandra.apache.org/doc/latest/cassandra/developing/cql/indexing/sai/sai-concepts.html) (SAI), which is much more powerful than Cassandra's current secondary index system, or Scylla's global/local secondary indexes. It allows for searching on indexed fields using vector search, AND/OR, IN, CONTAINS, numeric ranges, equality on numbers or strings, and more.

Are there any plans to add support for SAI to Scylla? It's a game-changing feature for many applications, and once Cassandra 5.0 reaches GA, I feel like many (myself included) would like to see the feature in Scylla as well.

SAI was previously mentioned by @nyh in https://github.com/scylladb/scylladb/issues/2203#issuecomment-1533238833 as the path forwards for adding Cassandra-compatible indexes to Scylla. This issue picks up where that comment left off, intending to serve as tracking for SAI support.

**Comment by avikivity** [comment]
SAI is anti-scalable. The more nodes and shards in your cluster, the more CPU it requires. It's even worse with tablets.

If SAI queries are a tiny fraction of overall work, then it works, but above a certain fraction it will dominate, and increasing the cluster size will not help.

**Comment by kostja** [comment]
@Saghetti0 I would be grateful if you talk a bit more about your use case. What kind of search would you like to speed up with better indexing? Could you provide example queries and describe their desired query plans?

**Comment by AhmedBHameed** [comment]
**I could be wrong due lack of knowledge.**

I am exploring Scylla DB recently and would like to see its potential compatibility. I just came from mongodb world where they have text_search via "$search" operator.

I was thinking about system logs where the user can search by logs.

I have something like this (Just simple example I implement)

```sql
CREATE TABLE IF NOT EXISTS xyz.system_logs
(
    system_logs_id uuid,
    day date,
    log text,
    created_at timestamp,

    PRIMARY KEY (day, created_at)
) WITH CLUSTERING ORDER BY (created_at DESC);
```

What I am doing currently is this

```sql
SELECT system_logs_id, day, log, created_at FROM xyz.system_logs
	WHERE day IN ?
    AND log LIKE ?
```

currently it is only working with "ALLOW FILTERING".

I was checking the secondary index but it is likely not supporting `LIKE ?`.

Not sure if that SAI will support text_search ?

**Comment by dorlaor** [comment]
Answer by Gemini3

Yes, ScyllaDB supports the `LIKE` operator, but with **major performance caveats** that differ from traditional SQL databases or Apache Cassandra.

### The Short Answer

* **Yes, it works:** You can use `LIKE` in your `WHERE` clause.
* **But:** It requires `ALLOW FILTERING` in almost all cases because it performs a "Full Table Scan" (or full partition scan). It does **not** use an index to speed up the search.

### 1. Syntax & Supported Wildcards

ScyllaDB's `LIKE` operator is actually more flexible than Cassandra's standard implementation in terms of pattern matching:

* **`%`**: Matches any string of zero or more characters. (Supported at the beginning, middle, or end).
* **`_`**: Matches any single character.
* **Case Sensitivity:** It is **case-sensitive**. `LIKE 'Scylla'` will not match "scylla".

**Example Query:**

```sql
SELECT * FROM users 
WHERE email LIKE '%@gmail.com' 
ALLOW FILTERING;

```

### 2. Critical Performance Warning

In ScyllaDB, `LIKE` is implemented as a **server-side filter**, not an indexed lookup.

* **How it works:** ScyllaDB reads the data from disk first, then checks if it matches your pattern, and throws away the rows that don't match.
* **The Cost:** If you run `SELECT * FROM table WHERE col LIKE '%val%'`, ScyllaDB must read **every single row in the table**. On a large dataset, this will cause timeouts and extreme load on the cluster.
* **When it is safe:** It is only safe to use if you restrict the query to a specific **Partition Key** first.
* *Safe:* `SELECT * FROM users WHERE user_id = 123 AND email LIKE 'A%'` (Scans only one user's data).
* *Dangerous:* `SELECT * FROM users WHERE email LIKE 'A%'` (Scans the entire database).



### 3. Comparison with Cassandra (SASI)

If you are coming from Cassandra, you might be looking for **SASI (SSTable Attached Secondary Index)**, which allowed indexed `LIKE` queries.

* **ScyllaDB does NOT support SASI.**
* The ScyllaDB team has stated they do not intend to implement SASI because it is considered unstable/experimental even in Cassandra and is being deprecated there.

### 4. Alternatives for Search in ScyllaDB

If you need performant text search (e.g., "Find all users whose name starts with 'Jo'"), `LIKE` is not the right tool. Use one of these approaches instead:

1. **Prefix Search (Clustering Keys):**
If you only need "Starts With" queries, model your table so the text column is a **Clustering Key**.
```sql
-- Efficient for 'Jo%' queries
SELECT * FROM users WHERE country='US' AND name >= 'Jo' AND name < 'Jp';

```


2. **External Indexing (Recommended):**
For true full-text search (contains, fuzzy match, case-insensitive), the standard practice is to dual-write data to **Elasticsearch** or use a dedicated search solution alongside ScyllaDB.
3. **Secondary Indexes (Exact Match Only):**
Standard ScyllaDB secondary indexes only support strict equality (`=`). They do not support `LIKE`.

**Comment by AhmedBHameed** [comment]
Thank you @dorlaor for the comment.

If I understand correctly

```text
When it is safe: It is only safe to use if you restrict the query to a specific Partition Key first.
Safe: SELECT * FROM users WHERE user_id = 123 AND email LIKE 'A%' (Scans only one user's data).
```
Then my query is fine with ALLOW FILTERING as long I have the partition key as day. It will search a subset of data and avoid full database search.

One other thing,
As the main reason makes me come to Scylla is that in your conference @ AppNexus, you mentioned that 2 clients removed Redis and just use the Scylla directly due to the speed. That was making me thinking about comprehensive solution in one place.

**Comment by dorlaor** [comment]
You're right, there is no need for an external cache with Scylla. We've done several blogs
and compared the results to Redis and memcache (with disk access).
